src/main/java/network/brightspots/rcv/DominionCvrReader.java

```diff
@@ -0,0 +1,264 @@
+    try {
+      HashMap json = JsonParser.readFromFile(filePath, HashMap.class);
+      // top-level "Sessions" object contains a lists of Cvr objects from different tabulators
+      ArrayList sessions = (ArrayList) json.get("Sessions");
+      // for each Cvr object extract various fields
+      for (Object sessionObject : sessions) {
+        HashMap session = (HashMap) sessionObject;
+
+        // extract various ids
+        Integer tabulatorId = (Integer) session.get("TabulatorId");
+        Integer batchId = (Integer) session.get("BatchId");
+        Integer recordId = (Integer) session.get("RecordId");
+        String suppliedId = recordId.toString();
+        HashMap originalObject = (HashMap) session.get("Original");
+
+        // validate precinct
+        Integer precinctPortionId = (Integer) originalObject.get("PrecinctPortionId");
+        if (!this.precincts.containsKey(precinctPortionId)) {
+          Logger.log(Level.SEVERE, "Precinct id %d from Cvr not found in manifest data!",
+              precinctPortionId);
+          throw new CvrParseException();
+        }
+        String precinct = this.precincts.get(precinctPortionId);
+        Integer ballotTypeId = (Integer) originalObject.get("BallotTypeId");
+
+        ArrayList contests;
+        // sometimes there is a "Cards" object at this level
+        if (originalObject.containsKey("Cards")) {
+          ArrayList cardsList = (ArrayList) originalObject.get("Cards");
+          HashMap cardsObject = (HashMap) cardsList.get(0);
+          contests = (ArrayList) cardsObject.get("Contests");
+        } else {
+          contests = (ArrayList) originalObject.get("Contests");
+        }
+
+        // each contest object is a cvr
+        for (Object contestObject : contests) {
+          HashMap contest = (HashMap) contestObject;
+          Integer contestId = (Integer) contest.get("Id");
+          // validate contest id
+          if (!this.contests.containsKey(contestId)) {
+            Logger.log(Level.SEVERE, "Contest id %d from Cvr not round in manifest data!",
+                contestId);
+            throw new CvrParseException();
+          }
+
+          ArrayList<Pair<Integer, String>> rankings = new ArrayList<>();
+          // marks is an array of rankings
+          ArrayList marks = (ArrayList) contest.get("Marks");
+          for (Object rankingObject : marks) {
+            HashMap rankingMap = (HashMap) rankingObject;
+            Integer candidateId = (Integer) rankingMap.get("CandidateId");
+            String candidateCode = candidateId.toString();
+            // validate candidate:
+            if (!this.contestIdToCandidateCodes.containsKey(contestId)) {
+              Logger.log(Level.SEVERE, "Unknown contest id %d found while parsing Cvr!", contestId);
+              throw new CvrParseException();
+            }
+            Set<String> candidates = this.contestIdToCandidateCodes.get(contestId);
```

Comment:
I think you can pull this outside of the loop (e.g. next to, or possibly even combined with, the check for `!this.contests.containsKey(contestId)`.

Reply:
dig it!